source/extensions/http/stateful_session/cookie/cookie.h

```diff
@@ -52,20 +62,55 @@ class CookieBasedSessionStateFactory : public Envoy::Http::SessionStateFactory {
   }
 
 private:
-  absl::optional<std::string> parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+  std::pair<absl::optional<std::string>, bool>
+  parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+    bool use_old_style_encoding = false;
     const std::string cookie_value = Envoy::Http::Utility::parseCookieValue(headers, name_);
-    std::string address = Envoy::Base64::decode(cookie_value);
+    if (cookie_value.empty()) {
+      return std::make_pair(absl::nullopt, use_old_style_encoding);
+    }
+    const std::string decoded_value = Envoy::Base64::decode(cookie_value);
+    std::string address;
+
+    // Try to interpret the cookie as proto payload.
+    // Otherwise treat it as "old" style format, which is ip-address:port.
+    envoy::Cookie cookie;
+    if (cookie.ParseFromString(decoded_value)) {
+      address = cookie.address();
+      if (address.empty() || (cookie.expires() == 0)) {
+        return std::make_pair(absl::nullopt, use_old_style_encoding);
+      }
+
+      std::chrono::seconds expiry_time(cookie.expires());
+      auto now = std::chrono::duration_cast<std::chrono::seconds>(
+          (time_source_.monotonicTime()).time_since_epoch());
+      if (now > expiry_time) {
+        // Ignore the address extracted from the cookie. This will cause
+        // upstream cluster to select a new host and new cookie will be generated.
+        return std::make_pair(absl::nullopt, use_old_style_encoding);
+      }
+    } else {
+      // Treat this as "old" style cookie.
+      ENVOY_LOG_MISC(
+          warn, "Non-proto cookie format detected. This format will be rejected in the future.");
```

Comment:
I think this log will bring lots of overhead to the envoy. If some users update the new version envoy, and their current users send some requests with old format cookie (a normal and predictable scenario), the envoy will be very very slow because a sync logging is used. I think maybe we need to ensure this log to print only once.

Reply:
Agree. Thanks for catching it. Indeed, during version upgrade Envoy may be bombarded with "old"-style cookies.